File: exercise1/k_nearest_neighbors.R
### Implementation of k-nearest neighbors algorithm
### tuanle@hotmail.de

## Algorithm: https://www.researchgate.net/figure/260397165_Pseudocode-for-KNN-classification

library(R6)

k_nearest_neighbors = R6Class("k_nearest_neighbors",
  public = list(
    X = NULL,
    y = NULL,
    n = NULL,
    p = NULL,
    g = NULL,
    nearest_neighbors = NULL,
    classification = FALSE,
    y_pred = NULL,
    ### Constructor ####
    # @param X [matrix(n,p)] or [data.frame(n,p)]
    #   Training matrix / dataframe for knn algorithm
    # @param y [numeric(n)] or [factor(n)] target vector depending on ml-problem. 
    #   returns k_nearest_neighbors instance
    initialize = function(X, y) {
      self$X = X
      self$y = y
      self$n = nrow(X)
      self$p = ncol(X)
      # if classification problem, get number of classes
      if (is.factor(y)) {
        self$g = length(levels(y))
        self$classification = TRUE
      } 
    },
    ### Train the non-parametric algorithm ###
    # @param k [integer(1)]. Default is to take 5 nearest neighbors
    #   Number of neighbors considered for classification/regression.
    # @return [invisible(NULL)]. 
    #   returns nothing. Results will be saved within k_nearest_neighbors instance.
    # for neighborhood check instance$nearest_neighbors
    # for prediction check instance$y_pred
    train_knn = function(k = 5, X_data = NULL, y_data = NULL){
      if (is.null(X_data) & is.null(y_data)) {
        X_data = self$X
        y_data = self$y
      }
      self$y_pred = rep(NA, times = nrow(X_data))
      self$nearest_neighbors = vector("list", length = nrow(X_data))
      # For each observation in X(_train) compute the distance to each other observation
      for (observation in seq.int(nrow(X_data))) {
        #out: nx1 vector, because n instances in train.
        dist = apply(X = self$X, MARGIN =  1, function(x) sum((x - X_data[observation, ])^2))
        # order by returning index of observations for distance in increasing order. 
        dist = order(dist, decreasing = FALSE)
        # take the k-nearest neighbors
        nearest_neighbors_idx = dist[1:k]
        # store into nearest_neighbors_list
        self$nearest_neighbors[[observation]] = nearest_neighbors_idx
        ### As of now code predict function within train
        if (self$classification) {
          ## If classification problem, make majority vote for all observation within a neighbourhood
          class_frequency_in_neighborhood = table(self$y[nearest_neighbors_idx])
          majority_vote = names(class_frequency_in_neighborhood)[which.max(class_frequency_in_neighborhood)]
          ## store prediction in y_train_pred vector
          self$y_pred[observation] = majority_vote
        } else {
          ## If regression, take the mean of the neighbor hood values
          mean_value_neighbors = mean(self$y[nearest_neighbors_idx])
          self$y_pred[observation] = mean_value_neighbors
        }
      }
      if (self$classification) {
        print(paste("Accuracy:", sum(y_data == self$y_pred)/nrow(X_data)))
      } else {
        print(paste("Accuracy (MSE):", sum((y_data - self$y_pred)^2)/nrow(X_data)))
      }
      return(invisible(NULL))
    },
    ### Predicts outcome for test data ###
    # @param X_test [matrix] or [data.frame] to predict target vector
    # @param y_test [numeric] or [factor] test target vector
    # @param k [integer(1)]  Number of neighbors considered for classification/regression.
    #   returns nothing. Results will be saved within k_nearest_neighbors instance.
    # for neighborhood check instance$nearest_neighbors
    # for prediction check instance$y_pred
    predict_knn = function(k = 5, X_test, y_test) {
      #call train_knn
      self$train_knn(k = k, X_data = X_test, y_data = y_test)
    }
  )
)

X = iris[, 1:4]
y = iris$Species

### Hold-Out-Sampling:
set.seed(1)
train_idx = sample(1:nrow(X), size = ceiling(2/3*nrow(X)), replace = FALSE)
test_idx = setdiff(1:nrow(X), train_idx)
X_train = X[train_idx,]
y_train = y[train_idx]
X_test = X[test_idx,]
y_test = y[test_idx]

knn_iris = k_nearest_neighbors$new(X = X_train, y = y_train)
## check init values
print(knn_iris$classification)
## train
knn_iris$train_knn(k = 5)
#[1] "Accuracy: 0.96"
print(knn_iris$y_pred)
## test
knn_iris$predict_knn(k = 5, X_test = X_test, y_test = y_test)
##[1] "Accuracy: 1"
sum(knn_iris$y_pred == y_test)/nrow(X_test)

## try regression
X = mtcars[,-1]
y = mtcars[,1]
### Hold-Out-Sampling:
set.seed(1)
train_idx = sample(1:nrow(X), size = ceiling(2/3*nrow(X)), replace = FALSE)
test_idx = setdiff(1:nrow(X), train_idx)
X_train = X[train_idx,]
y_train = y[train_idx]
X_test = X[test_idx,]
y_test = y[test_idx]

knn_cars = k_nearest_neighbors$new(X = X_train, y = y_train)
## check init values
print(knn_cars$classification)
## train
knn_cars$train_knn(k = 3)
#[1] "Accuracy (MSE): 5.42909090909091"
print(knn_cars$y_pred)
## test
knn_cars$predict_knn(k = 3, X_test = X_test, y_test = y_test)
#[1] "Accuracy (MSE): 2.24811111111111"
knn_cars$y_pred

## END

File: exercise1/k_nearest_neighbors.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: Tuan Le
@email: tuanle@hotmail.de

Implementation of k-nearest neighbours algorithm
"""

import numpy as np
from collections import Counter

class k_nearest_neighbors(object):
    
    ### Constructor ###
    def __init__(self, X, y, classif):
        '''
        Init method for K-Nearest-Neighbors algorithm.
        
        Parameters
        ----------
        X : Feature matrix
        y : Target vector
        classif : Boolean whether the algorithm should perform classification or regression.    
        Returns
        -------
        k_n_n object : an K-Nearest-Neighbours object for classification or regression
        '''
        self.X = X
        self.y = y
        self.n = X.shape[0]
        self.classif = classif
        if(classif):
            self.levels = np.unique(self.y)

    '''
    static method to compute euclidean distance
    '''
    @staticmethod
    def euclidean(xi, xj):
        di = np.sqrt(np.sum(np.square(xi - xj)))
        return di
    
    ### Train method ###
    def train(self, k = 5, X_data = None, y_data = None):
        ## Since prediction also calls train, we need to insert the *_data objects
        if X_data == None:
            X_data = self.X
        if y_data == None:
            y_data = self.y
        ## Init nearest_neighbors list
        self.nearest_neighbors = [None] * X_data.shape[0]
        ## Init pred vector for each observation
        self.y_pred = np.zeros(len(y_data))
        ## Now loop through each observation and compute the distance to all other points:
        for i in range(X_data.shape[0]):
            #for each iteration store distances from i to not_i in dist 
            dist = np.zeros(self.n)
            #dist = np.apply_along_axis(func1d=self.euclidean, axis=1, arr=X, xj = X[i:,])
            ## Inefficient way:
            for j in range(X_data.shape[0]):
                dist[j] = self.euclidean(xi = X_data[i,:], xj = X_data[j,:])
            ## Order distances and retrieve indexes
            #dist_ordered = np.sort(dist) ##for sorted
            ordered_idx = np.argsort(dist) ##for retrieving index of sorted array
            ## Take the k nearest neighbors
            k_nearest_neighbors = ordered_idx[0:k]
            ## Save results
            self.nearest_neighbors[i] = k_nearest_neighbors
            ## Predict
            if(self.classif):
                class_frequency_in_neighborhood = dict(Counter(y_data[k_nearest_neighbors]))
                ## apply majority vote, meaning getting the class which most neighbors have
                ## By default counter sorts in decreasing order, so we take the first key element from dict
                self.y_pred[i] = list(class_frequency_in_neighborhood.keys())[0]
            else:
                self.y_pred[i] = np.mean(a=y_data[k_nearest_neighbors])
        ## Now print performance measure:
        if(self.classif):
            acc = np.mean(y_data == self.y_pred)
            print("Accuracy in knn classification: {}".format(acc))
        else:
            mse = 1/len(y_data)* np.sum(np.square((y_data - self.y_pred)))
            print("Accuracy in knn classification: {}".format(mse))
        ##END train, return nothing
        return None
    
#    def predict(self, k = 5, X_test, y_test):
#        ## Call class train method
#        self.train(k, X_data = X_test, y_data = y_test)
#        return None
    
       
import sklearn.datasets as sdset
iris = sdset.load_iris()
X = iris["data"]
y = iris["target"]

my_knn = k_nearest_neighbors(X=X, y=y, classif=True)
my_knn.train(k=5)
#Accuracy in knn classification: 1.0